addons/account_audit_trail/models/res_partner.py

```diff
@@ -0,0 +1,19 @@
+from odoo import models
+
+from .mail_message import bypass_token
+
+
+class ResPartner(models.Model):
+    _inherit = "res.partner"
+
+    def unlink(self):
+        partner_with_audit_trail = self.filtered_domain([
+            '|', ('company_id', '=', False), ('company_id.check_account_audit_trail', '=', True),
+            '|', ('customer_rank', '>', 0), ('supplier_rank', '>', 0),
+        ])
+        if not partner_with_audit_trail:
+            # The domain used for filtering is a bit more broad than the records that need to be blocked, so if we
+            # don't have any here, we can bypass the heavy search on the partner's messages
+            return super(ResPartner, self.with_context(bypass_audit=bypass_token)).unlink()
```

Comment:
You should only set the token after having checked that there is no move linked to the partner by querying the `account_move` table, and this `if partner_with_audit_trail`

I can live with not being able to delete a message[^1] if we can at least delete the partner

Then you don't need any diff in `mail.message`, limiting the performance impact

[^1]: this could still be managed by segregating the records based on the model and do this in `_except_audit_log` instead of in this function

Reply:
I pushed a commit to reflect what I said in my footnote.
What do you think?